### PR TITLE
Support dynamic kubelet

### PR DESCRIPTION
**Pull request by bowang-666**
Translate current flags we set to config file before running kubelet.
Flags can be found in https://github.com/Azure/aks-engine/blob/aks-release-v0.47.0-1/pkg/api/defaults-kubelet.go
and some RP files.
Remove translated flags from flag string, but some flags will still remain (for example azure specific flags)
Configurations in config file can be changed at run time following https://kubernetes.io/docs/tasks/administer-cluster/reconfigure-kubelet/

### CODE REVIEW

**Comment by bowang-666** [comment]
going to add unit tests to verify translation

File: pkg/agent/utils.go [hunk 1]
```diff
@@ -808,3 +840,136 @@ func getCustomDataFromJSON(jsonStr string) string {
...
+	var buf bytes.Buffer
+	for _, key := range keys {
+		buf.WriteString(fmt.Sprintf("%s=%s ", key, k.KubeletConfig[key]))
+	}
+	return buf.String()
+}
+
+// IsDynamicKubeletSupported get if dynamic kubelet is supported in AKS
+func IsDynamicKubeletSupported(cs *api.ContainerService) bool {
+	return cs.Properties.OrchestratorProfile.IsKubernetes() && IsKubernetesVersionGe(cs.Properties.OrchestratorProfile.OrchestratorVersion, "1.14.0")
```
Comment (bowang-666) [inline pkg/agent/utils.go:864]: I assume we only care about new create clusters so I put 1.14.0 here for lowest support version.

File: parts/linux/cloud-init/nodecustomdata.yml [hunk 2]
```diff
@@ -344,11 +344,20 @@ write_files:
...
+    {{GetKubeletConfigFile .KubernetesConfig}}
+    #EOF
+{{end}}
+
 - path: /etc/default/kubelet
   permissions: "0644"
   owner: root
   content: |
-    KUBELET_CONFIG={{GetKubeletConfigKeyVals .KubernetesConfig }}
+    KUBELET_FLAGS={{GetKubeletConfigKeyVals .KubernetesConfig }}
```
Comment (xuto2) [inline parts/linux/cloud-init/nodecustomdata.yml:360]: Check L366

File: parts/linux/cloud-init/artifacts/kubelet.service [hunk 3]
```diff
@@ -32,8 +32,9 @@ ExecStart=/usr/local/bin/kubelet \
         --enable-server \
         --node-labels="${KUBELET_NODE_LABELS}" \
         --v=2 {{if NeedsContainerd}}--container-runtime=remote --runtime-request-timeout=15m --container-runtime-endpoint=unix:///run/containerd/containerd.sock{{end}} \
-        --volume-plugin-dir=/etc/kubernetes/volumeplugins \
-        $KUBELET_CONFIG \
+        --volume-plugin-dir=/etc/kubernetes/volumeplugins \{{if IsDynamicKubeletSupported}}
+        --config /etc/default/kubeletconfig.json)} \{{end}}
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/kubelet.service:36]: I just tried the hyphen syntax ({{- If ...}}) and it works. So you still put the if statement in a new line with a prefix hyphen

File: parts/linux/cloud-init/nodecustomdata.yml [hunk 4]
```diff
@@ -344,11 +344,20 @@ write_files:
     current-context: localclustercontext
     #EOF
 
+{{if IsDynamicKubeletSupported}}
+- path: /etc/default/kubeletconfig.json
+  permissions: "0644"
+  owner: root
+  content: |
+    {{GetKubeletConfigFile .KubernetesConfig}}
+    #EOF
```
Comment (xuto2) [inline parts/linux/cloud-init/nodecustomdata.yml:353]: remove this since

File: parts/linux/cloud-init/nodecustomdata.yml [hunk 5]
```diff
@@ -344,11 +344,20 @@ write_files:
...
+{{end}}
+
```
Comment (xuto2) [inline parts/linux/cloud-init/nodecustomdata.yml:355]: I guess you are moving this to cse_config?

File: pkg/agent/baker.go [hunk 6]
```diff
@@ -232,11 +232,20 @@ func getContainerServiceFuncMap(cs *api.ContainerService, profile *api.AgentPool
 		"GetAgentKubernetesLabelsDeprecated": func(profile *api.AgentPoolProfile, rg string) string {
 			return profile.GetKubernetesLabels(rg, true)
 		},
+		"GetKubeletConfigFile": func(kc *api.KubernetesConfig) string {
+			if kc == nil {
+				return ""
+			}
+			return getKubeletConfigFileFromFlags(kc.KubeletConfig)
```
Comment (xuto2) [inline pkg/agent/baker.go:239]: just use profile from L206

profile.KubernetesConfig.KubeletConfig

File: parts/linux/cloud-init/artifacts/kubelet.service (hunk 3 above)
Comment (bowang-666) [reply parts/linux/cloud-init/artifacts/kubelet.service]: changed to use {{- if

File: parts/linux/cloud-init/nodecustomdata.yml (hunk 5 above)
Comment (bowang-666) [reply parts/linux/cloud-init/nodecustomdata.yml]: correct

File: parts/linux/cloud-init/artifacts/kubelet.service [hunk 7]
```diff
@@ -33,7 +33,10 @@ ExecStart=/usr/local/bin/kubelet \
         --node-labels="${KUBELET_NODE_LABELS}" \
         --v=2 {{if NeedsContainerd}}--container-runtime=remote --runtime-request-timeout=15m --container-runtime-endpoint=unix:///run/containerd/containerd.sock{{end}} \
         --volume-plugin-dir=/etc/kubernetes/volumeplugins \
-        $KUBELET_CONFIG \
+        {{- if IsDynamicKubeletSupported}}
+        --config /etc/default/kubeletconfig.json --dynamic-config-dir /etc/default/dynamickubelet \
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/kubelet.service:37]: Just curious - do we need to pre-create the config dir or not?

File: pkg/agent/baker.go [hunk 8]
```diff
@@ -232,11 +227,20 @@ func getContainerServiceFuncMap(cs *api.ContainerService, profile *api.AgentPool
...
 			return profile.GetKubernetesLabels(rg, true)
 		},
+		"GetKubeletConfigFile": func() string {
+			if profile.KubernetesConfig == nil {
+				return ""
+			}
+			return getKubeletConfigFileFromFlags(profile.KubernetesConfig.KubeletConfig)
+		},
+		"IsDynamicKubeletSupported": func() bool {
+			return IsDynamicKubeletSupported(cs)
```
Comment (xuto2) [inline pkg/agent/baker.go:237]: I don't see the implementation of this function, does it exist already?

File: pkg/agent/junit.xml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (xuto2) [inline pkg/agent/junit.xml:1]: do we need this file?

File: pkg/agent/baker_test.go [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+		Entry("AKSUbuntu1604 with Temp Disk and containerd", "AKSUbuntu1604+TempDisk+Containerd", "1.15.7", func(cs *api.ContainerService) {
+			cs.Properties.OrchestratorProfile.KubernetesConfig.ContainerRuntimeConfig = map[string]string{
+				common.ContainerDataDirKey: "/mnt/containers",
+			}
+			cs.Properties.AgentPoolProfiles[0].KubernetesConfig.ContainerRuntime = api.Containerd
+		}),
+		Entry("AKSUbuntu1604 with RawUbuntu", "RawUbuntu", "1.15.7", func(cs *api.ContainerService) {
+			cs.Properties.AgentPoolProfiles[0].Distro = api.Ubuntu
+		}),
+		Entry("AKSUbuntu1604 with AKSCustomCloud", "AKSUbuntu1604+AKSCustomCloud", "1.15.7", func(cs *api.ContainerService) {
```
Comment (xuto2) [inline pkg/agent/baker_test.go:167]: Let's remove AKSCustomCloud related testing.

File: pkg/agent/baker_test.go [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+		}),
+		Entry("AKSUbuntu1604 with Temp Disk and containerd", "AKSUbuntu1604+TempDisk+Containerd", "1.15.7", func(cs *api.ContainerService) {
+			cs.Properties.OrchestratorProfile.KubernetesConfig.ContainerRuntimeConfig = map[string]string{
+				common.ContainerDataDirKey: "/mnt/containers",
+			}
+			cs.Properties.AgentPoolProfiles[0].KubernetesConfig.ContainerRuntime = api.Containerd
+		}),
+		Entry("AKSUbuntu1604 with RawUbuntu", "RawUbuntu", "1.15.7", func(cs *api.ContainerService) {
+			cs.Properties.AgentPoolProfiles[0].Distro = api.Ubuntu
+		}),
```
Comment (xuto2) [inline pkg/agent/baker_test.go:166]: I don't see a test case for dynamic kubelet enabled scenario?

File: pkg/agent/testdata/AKSUbuntu1604+AKSCustomCloud/CSECommand [hunk 12]
```diff
@@ -0,0 +1 @@
...ure-cni/v1.1.3/binaries/azure-vnet-cni-linux-amd64-v1.1.3.tgz CNI_PLUGINS_URL=https://acs-mirror.azureedge.net/cni/cni-plugins-amd64-v0.7.6.tgz CLOUDPROVIDER_BACKOFF=<nil> CLOUDPROVIDER_BACKOFF_MODE= CLOUDPROVIDER_BACKOFF_RETRIES=0 CLOUDPROVIDER_BACKOFF_EXPONENT=0 CLOUDPROVIDER_BACKOFF_DURATION=0 CLOUDPROVIDER_BACKOFF_JITTER=0 CLOUDPROVIDER_RATELIMIT=<nil> CLOUDPROVIDER_RATELIMIT_QPS=0 CLOUDPROVIDER_RATELIMIT_QPS_WRITE=0 CLOUDPROVIDER_RATELIMIT_BUCKET=0 CLOUDPROVIDER_RATELIMIT_BUCKET_WRITE=0 LOAD_BALANCER_DISABLE_OUTBOUND_SNAT=<nil> USE_MANAGED_IDENTITY_EXTENSION=false USE_INSTANCE_METADATA=false LOAD_BALANCER_SKU= EXCLUDE_MASTER_FROM_STANDARD_LB=true MAXIMUM_LOADBALANCER_RULE_COUNT=0 CONTAINER_RUNTIME= CONTAINERD_DOWNLOAD_URL_BASE=https://storage.googleapis.com/cri-containerd-release/ NETWORK_MODE= KUBE_BINARY_URL= USER_ASSIGNED_IDENTITY_ID=userAssignedID API_SERVER_NAME= IS_VHD=true GPU_NODE=false SGX_NODE=false AUDITD_ENABLED=false  CONFIG_GPU_DRIVER_IF_NEEDED=true ENABLE_GPU_DEVICE_PLUGIN_IF_NEEDED=true /usr/bin/nohup /bin/bash -c "/bin/bash /opt/azure/containers/provision.sh >> /var/log/azure/cluster-provision.log 2>&1" systemctl --no-pager -l status kubelet 2>&1 | head -n 100
```
Comment (xuto2) [inline pkg/agent/testdata/AKSUbuntu1604+AKSCustomCloud/CSECommand:1]: remove this test case

File: pkg/agent/testdata/AKSUbuntu1604+AKSCustomCloud/CustomData [hunk 13]
```diff
@@ -0,0 +1,169 @@
+[base64(concat('#cloud-config
+
```
Comment (xuto2) [inline pkg/agent/testdata/AKSUbuntu1604+AKSCustomCloud/CustomData:2]: remove this test case and files

File: pkg/agent/utils.go (hunk 1 above)
Comment (xuto2) [reply pkg/agent/utils.go:864]: can we get a flag from caller, like enableGPUDevicePlugin? We should be able to turn this off if something doesn't work as expected?

File: pkg/agent/utils.go [hunk 14]
```diff
@@ -808,3 +840,140 @@ func getCustomDataFromJSON(jsonStr string) string {
 	}
 	return customDataObj["customData"]
 }
+
+// GetOrderedKubeletConfigString returns an ordered string of key/val pairs
+// copied from AKS-Engine and filter out flags that already translated to config file
+func GetOrderedKubeletConfigString(k *api.KubernetesConfig, cs *api.ContainerService) string {
```
Comment (xuto2) [inline pkg/agent/utils.go:846]: nit: check if k != nil?

Comment (xuto2) [inline pkg/agent/utils.go:846]: How about GetOrderedKubeletConfig**Flag**String ?

File: pkg/agent/utils.go [hunk 15]
```diff
@@ -808,3 +840,140 @@ func getCustomDataFromJSON(jsonStr string) string {
...
+	return buf.String()
+}
+
+// IsDynamicKubeletSupported get if dynamic kubelet is supported in AKS
+func IsDynamicKubeletSupported(cs *api.ContainerService) bool {
+	return cs.Properties.OrchestratorProfile.IsKubernetes() && IsKubernetesVersionGe(cs.Properties.OrchestratorProfile.OrchestratorVersion, "1.14.0")
+}
+
+// convert kubelet flags we set to a file
+func getKubeletConfigFileFromFlags(kc map[string]string) string {
```
Comment (xuto2) [inline pkg/agent/utils.go:868]: nit: getKubeletConfigFileContent?

File: pkg/agent/utils.go [hunk 16]
```diff
@@ -25,6 +26,37 @@ import (
 	"github.com/pkg/errors"
 )
 
+var TranslatedKubeletConfigFlags map[string]bool = map[string]bool{
```
Comment (xuto2) [inline pkg/agent/utils.go:29]: nit: KubeletConfigFlagsInFile?

File: pkg/agent/utils.go (hunk 14 above)
Comment (xuto2) [reply pkg/agent/utils.go:846]: or GetOrderedKubelet**Static**ConfigString ?

File: pkg/agent/utils.go (hunk 15 above)
Comment (xuto2) [reply pkg/agent/utils.go:868]: getKubeletDynamicConfigFileContent?

**Comment by xuto2** [comment]
please add some e2e test and make it switchable between dynamic config and static config, otherwise lgtm

File: pkg/agent/utils.go (hunk 1 above)
Comment (xuto2) [reply pkg/agent/utils.go:864]: Can we hardcode this to false and merge this PR first? 
There are more open PR to agentbaker and I want the unit test change to be in ASAP so we can ask them to add unit test for their change

File: parts/linux/cloud-init/artifacts/kubelet.service (hunk 7 above)
Comment (bowang-666) [reply parts/linux/cloud-init/artifacts/kubelet.service:37]: we don't need to, the directory will be created if not exist

File: pkg/agent/junit.xml (hunk 9 above)
Comment (bowang-666) [reply pkg/agent/junit.xml]: removed

File: pkg/agent/baker_test.go (hunk 11 above)
Comment (bowang-666) [reply pkg/agent/baker_test.go:149]: will enable afterwards

File: pkg/agent/utils.go (hunk 14 above)
Comment (bowang-666) [reply pkg/agent/utils.go]: it was checked in baker.go but I think we should check kubeletconfig though

File: pkg/agent/utils.go (hunk 16 above)
Comment (bowang-666) [reply pkg/agent/utils.go:29]: I'll keep this one since I'd like to call out these are translated from flags

**Comment by bowang-666** [comment]
> please add some e2e test and make it switchable between dynamic config and static config, otherwise lgtm

will add e2e in RP

File: pkg/agent/utils.go (hunk 1 above)
Comment (bowang-666) [reply pkg/agent/utils.go]: hard coded as false and merge this PR for now

**Pull request merged**